File: monitor_cpp/src/MonitorApp.cpp
#include "MonitorApp.h"

#include <algorithm>
#include <sstream>

#include "DataTypes/Matrix44.h"

/**
 * @brief Constructor
 * @param target connection target in the following format: "hostname:port" or "ip:port", e.g. "localhost:5000"
 */
MonitorApp::MonitorApp(const std::string& target) : AppClient(std::string(APP_NAME), target) {}

/**
 * @brief Gets called on remote app function calls received from the robot control
 * @param function app function data
 */
void MonitorApp::AppFunctionHandler(const robotcontrolapp::AppFunction& function)
{}

/**
 * @brief Gets called on remote UI update requests received from the robot control
 * @param updates updated UI elements. Key is the element name, value contains the changes.
 */
void MonitorApp::UiUpdateHandler(const std::map<std::string, const robotcontrolapp::AppUIElement*>& updates)
{
    // Handle faster/slower buttons
    for (auto& update : updates)
    {
        if (update.second->state().state_case() == robotcontrolapp::AppUIElement::AppUIState::kButtonState)
        {
            bool isClicked = update.second->state().button_state() == robotcontrolapp::ButtonState::CLICKED;
            if (isClicked)
            {
                if (update.first == "buttonFaster")
                {
                    SetVelocity(std::min(100.0f, GetVelocity() + 10.0f));
                }
                else if (update.first == "buttonSlower")
                {
                    SetVelocity(std::max(0.0f, GetVelocity() - 10.0f));
                }
            }
        }
    }
}

/**
 * @brief Updates the system info UI
 */
void MonitorApp::UpdateSystemInfo() {
    App::DataTypes::SystemInfo info = GetSystemInfo();

    QueueSetText("textSoftwareVersion", info.version);
    switch (info.systemType)
    {
        default:
        case robotcontrolapp::SystemInfo_SystemType_Other:
            QueueSetText("textSystemType", "unknown");
            break;
        case robotcontrolapp::SystemInfo_SystemType_Linux_x86:
            QueueSetText("textSystemType", "Linux x86");
            break;
        case robotcontrolapp::SystemInfo_SystemType_Raspberry:
            QueueSetText("textSystemType", "Raspberry Pi");
            break;
        case robotcontrolapp::SystemInfo_SystemType_Windows:
            QueueSetText("textSystemType", "Windows");
            break;
    }
    QueueSetText("textProject", info.projectFile);
    QueueSetText("textProjectTitle", info.projectTitle);
    QueueSetText("textProjectAuthor", info.projectAuthor);
    QueueSetText("textRobot", info.robotType);
    QueueSetText("textVoltage", info.voltage == robotcontrolapp::SystemInfo_Voltage_Voltage48V ? "48V" : "24V");
    QueueSetText("textDeviceID", info.deviceID);
    QueueSetText("textRobotAxes", std::to_string(info.robotAxisCount));
    QueueSetText("textExternalAxes", std::to_string(info.externalAxisCount));
    QueueSetText("textToolAxes", std::to_string(info.toolAxisCount));
    QueueSetText("textPlatformAxes", std::to_string(info.platformAxisCount));
    QueueSetText("textDigitalIOModules", std::to_string(info.digitalIOModuleCount));
    QueueSetText("textCycleTarget", std::to_string(info.cycleTimeTarget) + " ms");
    QueueSetText("textCycleAvg", std::to_string(info.cycleTimeAverage) + " ms");
    QueueSetText("textCycleMin", std::to_string(info.cycleTimeMin) + " ms");
    QueueSetText("textCycleMax", std::to_string(info.cycleTimeMax) + " ms");
    QueueSetText("textWorkload", std::to_string(info.workload) + " %");
    SendQueuedUIUpdates();
}

/**
 * @brief Translates a referencing state to a human readable string
 * @param state 
 * @return 
 */
std::string translateReferencingState(App::DataTypes::RobotState::ReferencingState state) {
    switch (state)
    {
        default:
        case App::DataTypes::RobotState::ReferencingState::NOT_REFERENCED:
            return "not referenced";
        case App::DataTypes::RobotState::ReferencingState::IS_REFERENCED:
            return "referenced";
        case App::DataTypes::RobotState::ReferencingState::IS_REFERENCING:
            return "referencing...";
    }
}

std::string translateHardwareState(App::DataTypes::RobotState::HardwareState state) {
    std::stringstream ss;
    ss << std::hex << "0x" << ((int)state);
    return ss.str();
}

/**
 * @brief Updates the robot state UI
 * @param state new robot state
 */
void MonitorApp::UpdateRobotState(const App::DataTypes::RobotState& state)
{
    std::stringstream ssTcp;
    ssTcp << std::fixed << "X=" << state.tcp.GetX() << ", Y=" << state.tcp.GetY() << ", Z=" << state.tcp.GetZ() << ", A=" << state.tcp.GetA() << ", B=" << state.tcp.GetB()
          << ", C=" << state.tcp.GetC();
    QueueSetText("textTCPPosition",  ssTcp.str());
    
    QueueSetText("textA1Name", state.joints[0].name);
    QueueSetText("textA1PosTarget", std::to_string(state.joints[0].targetPosition));
    QueueSetText("textA1PosActual", std::to_string(state.joints[0].actualPosition));
    QueueSetText("textA1State", translateHardwareState(state.joints[0].hardwareState));
    QueueSetText("textA1Referencing", translateReferencingState(state.joints[0].referencingState));
    QueueSetText("textA1TempBoard", std::to_string(state.joints[0].temperatureBoard) + " °C");
    QueueSetText("textA1TempMotor", std::to_string(state.joints[0].temperatureMotor) + " °C");
    QueueSetText("textA1Current", std::to_string(state.joints[0].current) + " mA");
    
    QueueSetText("textE1Name", state.joints[6].name);
    QueueSetText("textE1PosTarget", std::to_string(state.joints[6].targetPosition));
    QueueSetText("textE1PosActual", std::to_string(state.joints[6].actualPosition));
    QueueSetText("textE1State", translateHardwareState(state.joints[6].hardwareState));
    QueueSetText("textE1Referencing", translateReferencingState(state.joints[6].referencingState));
    QueueSetText("textE1TempBoard", std::to_string(state.joints[6].temperatureBoard) + " °C");
    QueueSetText("textE1TempMotor", std::to_string(state.joints[6].temperatureMotor) + " °C");
    QueueSetText("textE1Current", std::to_string(state.joints[6].current) + " mA");
    QueueSetText("textE1Velocity", std::to_string(state.joints[6].targetVelocity));

    std::stringstream ssPlatform;
    ssPlatform << "X=" << state.platformX << ", Y=" << state.platformY << ", heading=" << state.platformHeading;
    QueueSetText("textPlatformPosition", ssPlatform.str());
    QueueSetText("textDIn21", state.digitalInputs[20] ? "High" : "Low");
    QueueSetText("textDOut21", state.digitalOutputs[20] ? "High" : "Low");
    QueueSetText("textGSig1", state.globalSignals[0] ? "High" : "Low");
    QueueSetText("textHWError", state.hardwareState);
    QueueSetText("textVelocityOverride", std::to_string(state.velocityOverride) + " %");
    QueueSetText("textCartVelocity", std::to_string(state.cartesianVelocity) + " mm/s");
    QueueSetText("textTempCPU", std::to_string(state.temperatureCPU) + " °C");
    QueueSetText("textSupplyVoltage", std::to_string(state.supplyVoltage) + " V");
    QueueSetText("textCurrentAll", std::to_string(state.currentAll) + " mA");
    QueueSetText("textReferencingState", translateReferencingState(state.referencingState));
    SendQueuedUIUpdates();
}

/**
 * @brief Updates the robot state UI
 */
void MonitorApp::UpdateRobotState()
{
    // You can also query the robot state like this, use this approach if you don't need frequent updates:
    App::DataTypes::RobotState state = GetRobotState();
    UpdateRobotState(state);
}

/**
 * @brief Is called when the robot state is updated (usually each 10 or 20ms). Override this method, start the stream by calling StartRobotStateStream().
 * @param state new robot state
 */
void MonitorApp::OnRobotStateUpdated(const App::DataTypes::RobotState& state) {
    // For automatic as-fast-as-possible updates override this method (OnRobotStateUpdated) and call StartRobotStateStream()
    UpdateRobotState(state);
}

File: monitor_python/MonitorApp.py
import sys
from AppClient import AppClient
from DataTypes import RobotState
from DataTypes.Matrix44 import Matrix44
from robotcontrolapp_pb2 import AppFunction, AppUIElement, ButtonState, CheckboxState, HardwareState, ReferencingState, SystemInfo
from google.protobuf.internal import containers as protobufContainers


class MonitorApp(AppClient):
    """This is an example app implementation"""
    
    def __init__(self, appName: str, target: str):
        """Initializes the app. Pass the app name (as defined in rcapp.xml) and socket to connect to (default: "localhost:5000")"""
        AppClient.__init__(self, appName, target)

    def _AppFunctionHandler(self, function: AppFunction):
        """Gets called on remote app function calls received from the robot control"""
        return

    def _UiUpdateHandler(self, updates: protobufContainers.RepeatedCompositeFieldContainer[AppUIElement]):
        """Gets called on remote UI update requests received from the robot control"""
        for update in updates:
            if update.state.HasField("button_state"):
                # Handle buttons
                isClicked = update.state.button_state == ButtonState.CLICKED
                if isClicked:
                    # Init
                    if update.element_name == "buttonFaster":
                        self.SetVelocityOverride(min(100, self.GetVelocityOverride() + 10))
                    if update.element_name == "buttonSlower":
                        self.SetVelocityOverride(max(0, self.GetVelocityOverride() - 10))

    def UpdateSystemInfo(self):
        """Updates the system info UI"""
        info = self.GetSystemInfo()

        self.QueueSetText("textSoftwareVersion", info.version)
        if info.systemType == SystemInfo.SystemType.Other:
            self.QueueSetText("textSystemType", "unknown")
        elif info.systemType == SystemInfo.SystemType.Linux_x86:
            self.QueueSetText("textSystemType", "Linux x86")
        elif info.systemType == SystemInfo.SystemType.Raspberry:
            self.QueueSetText("textSystemType", "Raspberry Pi")
        elif info.systemType == SystemInfo.SystemType.Windows:
            self.QueueSetText("textSystemType", "Windows")
        else:
            self.QueueSetText("textSystemType", "unknown")

        self.QueueSetText("textProject", info.projectFile)
        self.QueueSetText("textProjectTitle", info.projectTitle)
        self.QueueSetText("textProjectAuthor", info.projectAuthor)
        self.QueueSetText("textRobot", info.robotType)
        self.QueueSetText("textVoltage", str(info.voltage))
        self.QueueSetText("textDeviceID", info.deviceID)
        self.QueueSetText("textIsSimulation", str(info.isSimulation))
        self.QueueSetText("textRobotAxes", str(info.robotAxisCount))
        self.QueueSetText("textExternalAxes", str(info.externalAxisCount))
        self.QueueSetText("textToolAxes", str(info.toolAxisCount))
        self.QueueSetText("textPlatformAxes", str(info.platformAxisCount))
        self.QueueSetText("textDigitalIOModules", str(info.digitalIOModuleCount))
        self.QueueSetText("textCycleTarget", f"{info.cycleTimeTarget} ms")
        self.QueueSetText("textCycleAvg", f"{info.cycleTimeAverage} ms")
        self.QueueSetText("textCycleMin", f"{info.cycleTimeMin} ms")
        self.QueueSetText("textCycleMax", f"{info.cycleTimeMax} ms")
        self.QueueSetText("textWorkload", f"{info.workload} %")
        self.SendQueuedUIUpdates()
    
    def translateReferencingState(self, state: ReferencingState) -> str:
        """Translates a referencing state to a human readable string"""
        if state == ReferencingState.NOT_REFERENCED:
            return "not referenced"
        elif state == ReferencingState.IS_REFERENCED:
            return "referenced"
        elif state == ReferencingState.IS_REFERENCING:
            return "referencing..."
        else:
            return "n/a"
        
    def translateHardwareState(self, state: HardwareState) -> str:
        """Translates a hardware state to a human readable string"""
        return f"{state:#x}"
    
    def UpdateRobotState(self, state: RobotState):
        """Updates the robot state UI"""
        tcpStr = f"X={state.tcp.GetX()}, Y={state.tcp.GetY()}, Z={state.tcp.GetZ()}, A={state.tcp.GetA()}, B={state.tcp.GetB()}, C={state.tcp.GetC()}"
        self.QueueSetText("textTCPPosition", tcpStr)

        self.QueueSetText("textA1Name", state.joints[0].name)
        self.QueueSetText("textA1PosTarget", str(state.joints[0].targetPosition))
        self.QueueSetText("textA1PosActual", str(state.joints[0].actualPosition))
        self.QueueSetText("textA1State", self.translateHardwareState(state.joints[0].hardwareState))
        self.QueueSetText("textA1Referencing", self.translateReferencingState(state.joints[0].referencingState))
        self.QueueSetText("textA1TempBoard", f"{state.joints[0].temperatureBoard} °C")
        self.QueueSetText("textA1TempMotor", f"{state.joints[0].temperatureMotor} °C")
        self.QueueSetText("textA1Current", f"{state.joints[0].current} mA")

        self.QueueSetText("textE1Name", state.joints[6].name)
        self.QueueSetText("textE1PosTarget", str(state.joints[6].targetPosition))
        self.QueueSetText("textE1PosActual", str(state.joints[6].actualPosition))
        self.QueueSetText("textE1State", self.translateHardwareState(state.joints[6].hardwareState))
        self.QueueSetText("textE1Referencing", self.translateReferencingState(state.joints[6].referencingState))
        self.QueueSetText("textE1TempBoard", f"{state.joints[6].temperatureBoard} °C")
        self.QueueSetText("textE1TempMotor", f"{state.joints[6].temperatureMotor} °C")
        self.QueueSetText("textE1Current", f"{state.joints[6].current} mA")
        self.QueueSetText("textE1Velocity", str(state.joints[6].targetVelocity))

        platformStr = f"X={state.platformX}, Y={state.platformY}, Z={state.platformHeading}"
        self.QueueSetText("textPlatformPosition", platformStr)
        if state.digitalInputs[20]:
            self.QueueSetText("textDIn21", "High")    
        else:
            self.QueueSetText("textDIn21", "Low")    
        if state.digitalOutputs[20]:            
            self.QueueSetText("textDOut21", "High")
        else:
            self.QueueSetText("textDOut21", "Low")
        if state.globalSignals[0]:
            self.QueueSetText("textGSig1", "High")
        else:
            self.QueueSetText("textGSig1", "Low")
        self.QueueSetText("textHWError", state.hardwareState)
        self.QueueSetText("textVelocityOverride", f"{state.velocityOverride} %")
        self.QueueSetText("textCartVelocity", f"{state.cartesianVelocity} mm/s")
        self.QueueSetText("textTempCPU", f"{state.temperatureCPU} °C")
        self.QueueSetText("textSupplyVoltage", f"{state.supplyVoltage} V")
        self.QueueSetText("textCurrentAll", f"{state.currentAll} mA")
        self.QueueSetText("textReferencingState", self.translateReferencingState(state.referencingState))
        self.SendQueuedUIUpdates()
    
    def ReadAndUpdateRobotState(self):
        """Updates the robot state UI"""
        state = self.GetRobotState()
        self.UpdateRobotState(state)
    
    def OnRobotStateUpdated(self, state: RobotState):
        """Is called when the robot state is updated (usually each 10 or 20ms). Override this method, start the stream by calling StartRobotStateStream()."""
        self.UpdateRobotState(state)


